File: src/test/java/com/github/nsc/de/compiler/parser/MathTests.java
package com.github.nsc.de.compiler.parser;

import com.github.nsc.de.compiler.parser.node.expression.*;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.compiler.TestUtil.*;
import static com.github.nsc.de.compiler.parser.ParserTestUtil.*;

public class MathTests {

    @Test
    public void testBasicNumber() {

        IntegerNode node = parseSingle("<VariableUsageTest>", "10", IntegerNode.class);
        assertEquals(10, node.getNumber());

    }

    @Test
    public void testBasicExpr() {

        testBasic("10 + 3", 10, 3, AddNode.class);
        testBasic("10 - 3", 10, 3, SubNode.class);
    }

    @Test
    public void testBasicTerm() {

        testBasic("10 * 3", 10, 3, MulNode.class);
        testBasic("10 / 3", 10, 3, DivNode.class);
        testBasic("10 % 3", 10, 3, ModNode.class);
    }

    @Test
    public void testBasicPow() {

        testBasic("10 ^ 3", 10, 3, PowNode.class);
        testBasic("10 ** 3", 10, 3, PowNode.class);

    }

    @Test
    public void testPointBeforeLine() {

        AddNode add = parseSingle("<PointBeforeLineTest>", "1 + 2 * 3 ^ 4", AddNode.class);
        assertNotNull(add.getLeft());
        assertType(IntegerNode.class, add.getLeft());
        assertEquals(1, ((IntegerNode) add.getLeft()).getNumber());
        assertNotNull(add.getRight());
        assertType(MulNode.class, add.getRight());

        MulNode mul = (MulNode) add.getRight();
        assertNotNull(mul.getLeft());
        assertType(IntegerNode.class, mul.getLeft());
        assertEquals(2, ((IntegerNode) mul.getLeft()).getNumber());
        assertNotNull(mul.getRight());
        assertType(PowNode.class, mul.getRight());

        PowNode pow = (PowNode) mul.getRight();
        assertNotNull(pow.getLeft());
        assertType(IntegerNode.class, pow.getLeft());
        assertEquals(3, ((IntegerNode) pow.getLeft()).getNumber());
        assertNotNull(pow.getRight());
        assertType(IntegerNode.class, pow.getRight());
        assertEquals(4, ((IntegerNode) pow.getRight()).getNumber());

    }

    @Test
    public void testBrackets() {

        MulNode node = parseSingle("<BracketTest>", "2 * (4 + 3)", MulNode.class);
        assertNotNull(node.getLeft());
        assertType(IntegerNode.class, node.getLeft());
        assertEquals(2, ((IntegerNode) node.getLeft()).getNumber());
        assertNotNull(node.getRight());
        assertType(AddNode.class, node.getRight());

        AddNode add = (AddNode) node.getRight();
        assertNotNull(add.getLeft());
        assertType(IntegerNode.class, add.getLeft());
        assertEquals(4, ((IntegerNode) add.getLeft()).getNumber());
        assertNotNull(add.getRight());
        assertType(IntegerNode.class, add.getRight());
        assertEquals(3, ((IntegerNode) add.getRight()).getNumber());

    }

    private <T extends ExpressionNode> void testBasic(String input, double left, double right, Class<T> type) {

        T node = parseSingle('<'+type.getSimpleName().substring(type.getSimpleName().length() - 4)+"Test>", input, type);
        assertNotNull(node.getLeft());
        assertTrue(node.getLeft() instanceof DoubleNode || node.getLeft() instanceof IntegerNode);
        if(node.getLeft() instanceof DoubleNode) assertEquals(left, ((DoubleNode) node.getLeft()).getNumber());
        else assertEquals(left, ((IntegerNode) node.getLeft()).getNumber());
        assertNotNull(node.getRight());
        assertTrue(node.getRight() instanceof DoubleNode || node.getRight() instanceof IntegerNode);
        if(node.getRight() instanceof DoubleNode) assertEquals(right, ((DoubleNode) node.getRight()).getNumber());
        else assertEquals(right, ((IntegerNode) node.getRight()).getNumber());

    }
}

File: src/test/kotlin/com/github/nsc/de/shake/parser/MathTests.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import org.junit.jupiter.api.Test
import com.github.nsc.de.shake.parser.node.factor.IntegerNode
import com.github.nsc.de.shake.parser.node.factor.DoubleNode
import com.github.nsc.de.shake.parser.node.expression.*
import org.junit.jupiter.api.Assertions

class MathTests {
    @Test
    fun testBasicNumber() {
        val node = ParserTestUtil.parseSingle("<VariableUsageTest>", "10", IntegerNode::class.java)
        Assertions.assertEquals(10, node.number)
    }

    @Test
    fun testBasicExpr() {
        testBasic("10 + 3", 10.0, 3.0, AddNode::class.java)
        testBasic("10 - 3", 10.0, 3.0, SubNode::class.java)
    }

    @Test
    fun testBasicTerm() {
        testBasic("10 * 3", 10.0, 3.0, MulNode::class.java)
        testBasic("10 / 3", 10.0, 3.0, DivNode::class.java)
        testBasic("10 % 3", 10.0, 3.0, ModNode::class.java)
    }

    @Test
    fun testBasicPow() {
        testBasic("10 ** 3", 10.0, 3.0, PowNode::class.java)
    }

    @Test
    fun testPointBeforeLine() {
        val add = ParserTestUtil.parseSingle("<PointBeforeLineTest>", "1 + 2 * 3 ** 4", AddNode::class.java)
        Assertions.assertNotNull(add.left)
        assertType(IntegerNode::class.java, add.left)
        Assertions.assertEquals(1, (add.left as IntegerNode).number)
        Assertions.assertNotNull(add.right)
        assertType(MulNode::class.java, add.right)
        val mul = add.right as MulNode
        Assertions.assertNotNull(mul.left)
        assertType(IntegerNode::class.java, mul.left)
        Assertions.assertEquals(2, (mul.left as IntegerNode).number)
        Assertions.assertNotNull(mul.right)
        assertType(PowNode::class.java, mul.right)
        val pow = mul.right as PowNode
        Assertions.assertNotNull(pow.left)
        assertType(IntegerNode::class.java, pow.left)
        Assertions.assertEquals(3, (pow.left as IntegerNode).number)
        Assertions.assertNotNull(pow.right)
        assertType(IntegerNode::class.java, pow.right)
        Assertions.assertEquals(4, (pow.right as IntegerNode).number)
    }

    @Test
    fun testBrackets() {
        val node = ParserTestUtil.parseSingle("<BracketTest>", "2 * (4 + 3)", MulNode::class.java)
        Assertions.assertNotNull(node.left)
        assertType(IntegerNode::class.java, node.left)
        Assertions.assertEquals(2, (node.left as IntegerNode).number)
        Assertions.assertNotNull(node.right)
        assertType(AddNode::class.java, node.right)
        val add = node.right as AddNode
        Assertions.assertNotNull(add.left)
        assertType(IntegerNode::class.java, add.left)
        Assertions.assertEquals(4, (add.left as IntegerNode).number)
        Assertions.assertNotNull(add.right)
        assertType(IntegerNode::class.java, add.right)
        Assertions.assertEquals(3, (add.right as IntegerNode).number)
    }

    private fun <T : ExpressionNode?> testBasic(input: String, left: Double, right: Double, type: Class<T>) {
        val node = ParserTestUtil.parseSingle(
            '<'.toString() + type.simpleName.substring(type.simpleName.length - 4) + "Test>",
            input,
            type
        )
        Assertions.assertNotNull(node!!.left)
        Assertions.assertTrue(node.left is DoubleNode || node.left is IntegerNode)
        if (node.left is DoubleNode) Assertions.assertEquals(
            left,
            (node.left as DoubleNode).number
        ) else Assertions.assertEquals(left, (node.left as IntegerNode).number.toDouble())
        Assertions.assertNotNull(node.right)
        Assertions.assertTrue(node.right is DoubleNode || node.right is IntegerNode)
        if (node.right is DoubleNode) Assertions.assertEquals(
            right,
            (node.right as DoubleNode).number
        ) else Assertions.assertEquals(right, (node.right as IntegerNode).number.toDouble())
    }
}